Hola Rafael!

Mi nombre es Matias y voy a estar revisando tu proyecto.

En la primer revision, de encontrar errores voy a estar marcandotelos para que puedas corregirlos por tu cuenta. Lo mas parecido a una situacion real de trabajo como DA. De haber posteriores revisiones

Encontraras mis comentarios debajo - **por favor no los muevas, modifiques o elimines**.

Encontraras mis comentarios en verde, amarillo o rojo de esta manera:

<div class="alert alert-block alert-success">
<b>Comentario de Revisor</b> <a class="tocSkip"></a>

Genial, buen trabajo.
</div>

<div class="alert alert-block alert-warning">
<b>Comentario de Revisor</b> <a class="tocSkip"></a>

Cuidado, se recomienda...
</div>

<div class="alert alert-block alert-danger">

<b>Comentario de Revisor</b> <a class="tocSkip"></a>

Necesita correcion. El trabajo no puede ser aceptado si precisa correcion.
</div>

Puedes responder a mis comentarios utilizando un mensaje de este tipo:

<div class="alert alert-block alert-info">
<b>Respuesta de estudiante.</b> <a class="tocSkip"></a>
</div>

# Introduccion al machine learning

## Introducción

Descripción del proyecto
La compañía móvil Megaline no está satisfecha al ver que muchos de sus clientes utilizan planes heredados. Quieren desarrollar un modelo que pueda analizar el comportamiento de los clientes y recomendar uno de los nuevos planes de Megaline: Smart o Ultra.

Tienes acceso a los datos de comportamiento de los suscriptores que ya se han cambiado a los planes nuevos (del proyecto del curso de Análisis estadístico de datos). Para esta tarea de clasificación debes crear un modelo que escoja el plan correcto. Como ya hiciste el paso de procesar los datos, puedes lanzarte directo a crear el modelo.

Desarrolla un modelo con la mayor exactitud posible. En este proyecto, el umbral de exactitud es 0.75. Usa el dataset para comprobar la exactitud.

**Instrucciones del proyecto.**  
1. Abre y examina el archivo de datos. Dirección al archivo: ***datasets/users_behavior.csv*** **Descarga el dataset**
2. Segmenta los datos fuente en un conjunto de entrenamiento, uno de validación y uno de prueba.
3. Investiga la calidad de diferentes modelos cambiando los hiperparámetros. Describe brevemente los hallazgos del estudio.
4. Comprueba la calidad del modelo usando el conjunto de prueba.
5. Tarea adicional: haz una prueba de cordura al modelo. Estos datos son más complejos que los que habías usado antes así que no será una tarea fácil. Más adelante lo veremos con más detalle.

**Descripción de datos**  
Cada observación en el dataset contiene información del comportamiento mensual sobre un usuario. La información dada es la siguiente:    
* **сalls** — número de llamadas,
* **minutes** — duración total de la llamada en minutos,
* **messages** — número de mensajes de texto,
* **mb_used** — Tráfico de Internet utilizado en MB,
* **is_ultra** — plan para el mes actual (Ultra - 1, Smart - 0).  

**Evaluación del proyecto**  
Hemos definido los criterios de evaluación para el proyecto. Lee esto con atención antes de pasar al ejercicio.  
Esto es lo que los revisores buscarán cuando evalúen tu proyecto:  
¿Cómo leíste los datos después de descargarlos?  
¿Segmentaste correctamente los datos en conjuntos de entrenamiento, validación y prueba?  
¿Cómo escogiste el tamaño de los conjuntos?  
¿Evaluaste correctamente la calidad del modelo?  
¿Qué modelos e hiperparámentros usaste?  
¿Cuáles fueron tus hallazgos?  
¿Probaste los modelos correctamente?  
¿Cuál es tu puntuación de exactitud?  
¿Te ceñiste a la estructura del proyecto y mantuviste limpio el código?  
Tienes tus hojas informativas y los resúmenes de los capítulos así que ya puedes continuar con el proyecto.  
¡Buena suerte!  

## Exploracion de datos

### Librerias a utilizar

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split

### Carga de datos

In [2]:
df = pd.read_csv('/datasets/users_behavior.csv')

### Verficación de datos

In [3]:
def display_df(data):
    data.info()
    display(data.head())
    print(data.isna().sum())
    print(data.describe())

In [4]:
display_df(df)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0

calls       0
minutes     0
messages    0
mb_used     0
is_ultra    0
dtype: int64
             calls      minutes     messages       mb_used     is_ultra
count  3214.000000  3214.000000  3214.000000   3214.000000  3214.000000
mean     63.038892   438.208787    38.281269  17207.673836     0.306472
std      33.236368   234.569872    36.148326   7570.968246     0.461100
min       0.000000     0.000000     0.000000      0.000000     0.000000
25%      40.000000   274.575000     9.000000  12491.902500     0.000000
50%      62.000000   430.600000    30.000000  16943.235000     0.000000
75%      82.000000   571.927500    57.000000  21424.700000     1.000000
max     244.000000  1632.060000   224.000000  49745.730000     1.000000


Observamos que los tipos de datos estan correctos y no contamos con valores ausentes, tenemos 3214 entradas de datos

<div class="alert alert-block alert-info">
<b>Respuesta de estudiante.</b> <a class="tocSkip"></a>  

Corregido
</div>

<div class="alert alert-block alert-success">
<b>Comentario de Revisor</b> <a class="tocSkip"></a>

Genial, buen comienzo.
</div>

<div class="alert alert-block alert-warning">
<b>Comentario de Revisor</b> <a class="tocSkip"></a>

A la funcion display_df que creaste seria bueno sumarle una linea mas que haga uso del metodo describe() de pandas para ver la distribucion de las variables numericas que tenga el dataset
</div>

## Segmentación de datos

In [5]:
df_train, df_rest = train_test_split(df, test_size = 0.40, random_state = 92)

In [6]:
df_valid, df_test = train_test_split(df_rest, test_size = 0.5, random_state = 92) 

In [7]:
print(f'Dimensiones del conjunto de entranamiento: {df_train.shape} en % respecto al conjunto original {(len(df_train)/len(df)):.0%}')
print(f'Dimensiones del conjunto de validación: {df_valid.shape} en % respecto al conjunto original {(len(df_valid)/len(df)):.0%}')
print(f'Dimensiones del conjunto de entranamiento: {df_test.shape} en % respecto al conjunto original {(len(df_test)/len(df)):.0%}')

Dimensiones del conjunto de entranamiento: (1928, 5) en % respecto al conjunto original 60%
Dimensiones del conjunto de validación: (643, 5) en % respecto al conjunto original 20%
Dimensiones del conjunto de entranamiento: (643, 5) en % respecto al conjunto original 20%


Separamos el conjunto de datos original en 3 partes:  
* Conjunto de entramiento
* Conjunto de validación
* Conjunto de prueba  

Se dividio con una relacion 3:1:1 es decir: el 60%, 20%, 20% respectivamente y lo podemos afirmar con la comprobación  

Como bien sabemos, nuestro objetivo es recomendar el tipo de plan, por lo que dividiremos nuestros datos por sus caracteristicas/features y por su objetivo/target que en este caso seria el plan la cual su columna es **'is_ultra'**.

In [8]:
# Segmentacion de datos para el conjunto de entramiento
train_features = df_train.drop(['is_ultra'], axis=1)
train_target = df_train['is_ultra']

In [9]:
# Segmentacion de datos para el conjunto de validación
valid_features = df_valid.drop(['is_ultra'], axis=1)
valid_target = df_valid['is_ultra']

In [10]:
# Segmentacion de datos para el conjunto de prueba
test_features = df_test.drop(['is_ultra'], axis=1)
test_target = df_test['is_ultra']

Ya tenemos nuestras caracteristicas y nuestros objetivos para cada tipo de conjunto, pasaremos a realizar los modelos y ponerlos a prueba

<div class="alert alert-block alert-success">
<b>Comentario de Revisor</b> <a class="tocSkip"></a>

Genial, buen trabajo con el split y separando las features de la variable objetivo
</div>

## Modelado de datos

El objetivo es recomendar un plan y en este caso tenemos plan smart(0) o ultra(1), por lo que tenemos un problema de clasificación, asi que haremos los modelos correspondientes para verificar cual es el más adecuado 

In [11]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
from sklearn.model_selection import GridSearchCV

### Arbol de decisión - Tree decision classifier

In [12]:
tree_model = DecisionTreeClassifier(random_state=92)

In [13]:
tree_param = {
    'max_depth' : [None, 10, 50, 100],
    'min_samples_leaf' : [1,2]
    }

In [14]:
gs1 = GridSearchCV(estimator = tree_model,
                  param_grid = tree_param,
                  cv = 5)

In [15]:
gs1.fit(train_features, train_target)

GridSearchCV(cv=5, estimator=DecisionTreeClassifier(random_state=92),
             param_grid={'max_depth': [None, 10, 50, 100],
                         'min_samples_leaf': [1, 2]})

In [16]:
print('Mejor modelo:',gs1.best_estimator_)
print()
print('Mejores hiperparametros:',gs1.best_params_)
print()
print('Mejor puntuación',gs1.best_score_)
print()

Mejor modelo: DecisionTreeClassifier(max_depth=10, min_samples_leaf=2, random_state=92)

Mejores hiperparametros: {'max_depth': 10, 'min_samples_leaf': 2}

Mejor puntuación 0.7806096494179396



In [17]:
best_tree_model = gs1.best_estimator_

In [18]:
train_score = best_tree_model.score(train_features, train_target)
valid_score = best_tree_model.score(valid_features, valid_target)

In [19]:
print('Puntuación del conjunto de entrenamiento de acuerdo al mejor modelo de arbol de decisión:',train_score)
print('Puntuación del conjunto de validación de acuerdo al mejor modelo de arbol de decisión:',valid_score)

Puntuación del conjunto de entrenamiento de acuerdo al mejor modelo de arbol de decisión: 0.870850622406639
Puntuación del conjunto de validación de acuerdo al mejor modelo de arbol de decisión: 0.7993779160186625


Realizamos la prueba de score al conjunto de entramiento y de validación para saber si tenemos overfitting, a lo cual vemos que la puntuacion del conjunto de validacion si es mas baja en relación a la puntuacion del conjunto de entramiento alrededor de 8% o 0.08.

### Bosque aleatorio - Random Forest Classifier

In [20]:
random_model = RandomForestClassifier(random_state=92)

In [21]:
random_param = {
    'max_depth' : [None, 10, 50, 100],
    'n_estimators' : [100,200,500],
    'min_samples_leaf' : [1,2]
    }

In [22]:
gs2 = GridSearchCV(estimator = random_model,
                  param_grid = random_param,
                  cv = 5)

In [23]:
gs2.fit(train_features, train_target)

GridSearchCV(cv=5, estimator=RandomForestClassifier(random_state=92),
             param_grid={'max_depth': [None, 10, 50, 100],
                         'min_samples_leaf': [1, 2],
                         'n_estimators': [100, 200, 500]})

In [24]:
print('Mejor modelo de bosque aleatorio:',gs2.best_estimator_)
print()
print('Mejores hiperparametros:',gs2.best_params_)
print()
print('Mejor puntuación:',gs2.best_score_)


Mejor modelo de bosque aleatorio: RandomForestClassifier(max_depth=10, n_estimators=500, random_state=92)

Mejores hiperparametros: {'max_depth': 10, 'min_samples_leaf': 1, 'n_estimators': 500}

Mejor puntuación: 0.8137971872686898


In [25]:
best_forest_model = gs2.best_estimator_

In [26]:
train_score = best_forest_model.score(train_features, train_target)
valid_score = best_forest_model.score(valid_features, valid_target)

In [27]:
print('Puntuación del conjunto de entrenamiento de acuerdo al mejor modelo de bosque aleatorio:',train_score)
print('Puntuación del conjunto de validación de acuerdo al mejor modelo de bosque aleatorio:',valid_score)

Puntuación del conjunto de entrenamiento de acuerdo al mejor modelo de bosque aleatorio: 0.899896265560166
Puntuación del conjunto de validación de acuerdo al mejor modelo de bosque aleatorio: 0.8164852255054432


Realizamos la prueba de score al conjunto de entramiento y de validación para saber si tenemos overfitting, a lo cual vemos que la puntuacion del conjunto de validacion es mas baja en relación a la puntuacion del conjunto de entramiento pero solo de 8% o 0.08.

### Regresion logistica - Logistic Regression

In [28]:
log_reg_model = LogisticRegression(random_state=92,solver='liblinear')

In [29]:
log_reg_param = {
    'penalty' : ['l1', 'l2'],
    'C' : [0.5, 1.0, 1.5, 2],
    }

In [30]:
gs3 = GridSearchCV(estimator = log_reg_model,
                  param_grid = log_reg_param,
                  cv = 5)

In [31]:
gs3.fit(train_features, train_target)

GridSearchCV(cv=5,
             estimator=LogisticRegression(random_state=92, solver='liblinear'),
             param_grid={'C': [0.5, 1.0, 1.5, 2], 'penalty': ['l1', 'l2']})

In [32]:
print('Mejor modelo de regresión logistica:',gs3.best_estimator_)
print()
print('Mejores hiperparametros:',gs3.best_params_)
print()
print('Mejor puntuación:',gs3.best_score_)


Mejor modelo de regresión logistica: LogisticRegression(C=0.5, penalty='l1', random_state=92, solver='liblinear')

Mejores hiperparametros: {'C': 0.5, 'penalty': 'l1'}

Mejor puntuación: 0.7541578628625261


In [33]:
best_log_reg_model = gs3.best_estimator_

In [34]:
train_score = best_log_reg_model.score(train_features, train_target)
valid_score = best_log_reg_model.score(valid_features, valid_target)

In [35]:
print('Puntuación del conjunto de entrenamiento de acuerdo al mejor modelo de bosque aleatorio:',train_score)
print('Puntuación del conjunto de validación de acuerdo al mejor modelo de bosque aleatorio:',valid_score)

Puntuación del conjunto de entrenamiento de acuerdo al mejor modelo de bosque aleatorio: 0.7567427385892116
Puntuación del conjunto de validación de acuerdo al mejor modelo de bosque aleatorio: 0.7340590979782271


Realizamos la prueba de score al conjunto de entramiento y de validación para saber si tenemos overfitting, a lo cual vemos que la puntuacion del conjunto de validacion es mas baja en relación a la puntuacion del conjunto de entramiento pero solo de 2% o 0.02.

### Conclusión de modelos

Podemos ver que el modelo que tiene una mayor puntiación/exactitud es el modelo de bosque aleatorio: RandomForestClassifier(max_depth=10, n_estimators=500, random_state=92)  

Con una exactitud de:  
0.899896265560166 para el conjunto de entramiento   
0.8164852255054432 para el conjunto de validación

Por lo que usaremos este modelo para hacer las predicciones con el conjunto de prueba

<div class="alert alert-block alert-info">
<b>Respuesta de estudiante.</b> <a class="tocSkip"></a>  

Corregido e implementado, se implemento GridSearchCV, es más practico a la hora de probar diferentes hiperparametros y mas sencillo de aplicar.
</div>

<div class="alert alert-block alert-success">
<b>Comentario de Revisor</b> <a class="tocSkip"></a>

Veo que has entendido la logica detras de entrenar los modelos con diferentes hiperparametros y corroborar cual es la mejor opcion en funcion de la metrica accuracy
    
Hiciste un muy buen trabajo
</div>

<div class="alert alert-block alert-danger">
<b>Comentario de Revisor</b> <a class="tocSkip"></a>

Rafael, en este seccion te propongo 2 correcciones
    
Primero, para elegir cuales hiperparametros serian optimos esta bien hacerlo con una metrica sobre las predicciones en el split de validacion. 
    
Pero tambien seria bueno que luego de haber encontrado el valor optimo de hiperparametros esa metrica la calcules tambien para el split train y test para comparar y concluir si estamos teniendo overfitting o no
    
Segundo, en vez de usar for loops, podrias usar el metodo Grid Search CV de la libreria sklearn y asi no solo probar con diferentes valores de 1 hiperparametro sino de 2 o mas hiperparametros a la vez.
    
Por ejemplo para
    
    - el arbol de decision con los hiperparametros max_depth y min_samples_leaf 
    
    - el random forest con max_depth y n_estimators
    
    - la regresion logistica con penalty y C
    
Eso si, al usar este metodo no uses listas de valores de hipermarametros muy grandes porque sino se pondra muy lento el entrenamiento (situacion que en la realidad se traduce en un alto costo de computo)
</div>



## Puesta a prueba

In [36]:
test_predict = best_forest_model.predict(test_features)

In [37]:
print(test_predict)

[1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 1 0 0 0 0 0 1 0 0 1 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 1 0 0 0 1 0 0 0 1 1 0 0 0 0 0 0 0 1 1 0
 1 0 0 0 1 0 0 0 0 0 0 1 0 1 0 0 0 0 0 1 0 0 0 0 0 0 1 1 0 0 1 0 0 1 0 0 0
 0 0 0 1 1 1 1 0 0 0 1 0 0 0 1 0 0 0 1 1 1 0 1 1 0 0 0 0 0 1 0 1 0 0 1 0 0
 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 1 0 1 0 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0
 0 1 0 0 0 1 0 0 1 1 0 0 0 0 0 1 0 0 1 0 1 0 0 0 1 0 1 0 0 1 0 0 1 0 1 1 0
 0 0 0 0 0 0 0 0 0 1 1 0 0 0 0 0 0 1 0 0 0 0 1 0 0 0 0 0 0 0 1 0 0 1 0 0 1
 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 1 1 1 0 1 0
 1 0 0 1 1 0 0 0 0 1 0 0 0 0 1 0 1 0 0 1 0 0 0 0 0 0 0 1 0 0 0 1 0 0 0 0 0
 1 1 0 1 1 0 0 0 1 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 1 1 1 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 1 0 0 1 1 0 0 1 0 0 0 0 0 0 0 0 0 0
 0 0 1 1 0 1 1 1 0 0 0 0 0 0 1 0 0 0 0 0 0 1 1 0 0 0 0 0 0 0 1 0 0 1 1 0 0
 0 0 0 0 0 0 0 0 0 1 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 1 0 0 0
 1 0 0 0 0 0 0 1 1 0 1 0 

In [38]:
test_accu = accuracy_score(test_target, test_predict)
test_precision = precision_score(test_target, test_predict)
test_recall = recall_score(test_target, test_predict)
print(f'La exactitud de la prediccion es de {test_accu}')
print(f'La precision del modelo es de {test_precision}')
print(f'La sensibilidad del modelo es de {test_recall}')

La exactitud de la prediccion es de 0.7931570762052877
La precision del modelo es de 0.7426470588235294
La sensibilidad del modelo es de 0.507537688442211


## Conclusión

De acuerdo a nuestro modelo la prediccion de los datos del conjunto de prueba es del 79%  
La precision nos indica el porcentaje que el modelo tomo como 1 en este caso seria el plan ultra respecto a los valores originales, lo cual nos da un 74.2% de precisión.  
La sensibilidad nos indica el porcentaje que el modelo reconocio de los planes que estaban marcados como 1 plan ultra, que en este caso fue de 50.75%

<div class="alert alert-block alert-info">
<b>Respuesta de estudiante.</b> <a class="tocSkip"></a>  

Corregido de hecho al final, las puntuaciones finales si variaron para bien, aunque no fue mucho, puntos son puntos.
</div>

<div class="alert alert-block alert-success">
<b>Comentario de Revisor</b> <a class="tocSkip"></a>

Rafael, tu proyecto esta muy bien encarado es claro que comprendiste los conceptos y los aplicaste con buen codigo
</div>

<div class="alert alert-block alert-danger">
<b>Comentario de Revisor</b> <a class="tocSkip"></a>

Las correcciones propuestas son minimas y espero te lleven a ahondar un poco mas sobre herramientas de ML
    
Quedo a la espera de tu devolucion con estas para poder aprobar tu proyecto, vas por muy buen camino
</div>

<div class="alert alert-block alert-success">
<b>Comentario de Revisor - 2da ITERACION</b> <a class="tocSkip"></a>

Excelente trabajo Rafel, felicitaciones tu proyecto esta aprobado
    
Que bueno que hayas encontrado provechoso el uso de Grid Search y hayas podido implementar todas las correcciones propuestas
</div>